modules/compute_statistics.nf

```diff
@@ -0,0 +1,160 @@
+
+process gc_bias {
+    cpus 1
+    tag { library }
+    conda "bioconda::picard=2.20.7"
+    publishDir "${params.flowcell}/${library}/stats/gc_bias"
+
+    input:
+        tuple val(library), path(bam), path(bai), val(barcodes)
+
+    output:
+        tuple val(params.email), val(library), path('*gc_metrics'), emit: for_agg
+
+    shell:
+    '''
+    picard -Xmx4g CollectGcBiasMetrics IS_BISULFITE_SEQUENCED=true VALIDATION_STRINGENCY=SILENT I=!{bam} O=!{library}.gc_metrics S=!{library}.gc_summary_metrics CHART=!{library}.gc.pdf R=!{params.genome}
+    '''
+}
+
+process idx_stats {
+    label 'cpus_8'
+    tag { library }
+    conda "bioconda::samtools=1.9"
+    publishDir "${params.flowcell}/${library}/stats/idxstats"
+
+    input:
+        tuple val(library), path(bam), path(bai), val(barcodes)
+
+    output:
+        tuple val(params.email), val(library), path("*idxstat"), emit: for_agg
+
+    shell:
+    '''
+    samtools idxstats !{bam} > !{library}.idxstat
+    '''
+}
+
+process flag_stats {
+    label 'cpus_8'
+    tag { library }
+    conda "bioconda::samtools=1.9"
+    publishDir "${params.flowcell}/${library}/stats/flagstats"
+
+    input:
+        tuple val(library), path(bam), path(bai), val(barcodes)
+
+    output:
+        tuple val(params.email), val(library), path("*flagstat"), emit: for_agg
+
+    shell:
+    '''
+    samtools flagstat -@!{task.cpus} !{bam} > !{library}.flagstat
+    '''
+}
+
+process fast_qc {
+    cpus 1
+    tag { library }
+    conda "bioconda::fastqc=0.11.8"
+    publishDir "${params.flowcell}/${library}/stats/fastqc"
+
+    input:
+        tuple val(library), path(bam), path(bai), val(barcodes)
+
+    output:
+        tuple val(params.email), val(library), path('*_fastqc.zip'), emit: for_agg
+
+    shell:
+    '''
+    fastqc -f bam !{bam}
+    '''
+}
+
+process insert_size_metrics {
+    cpus 1
+    tag { library }
+    conda "bioconda::picard=2.20.7 bioconda::samtools=1.9"
+    publishDir "${params.flowcell}/${library}/stats/insert_size"
+
+    input:
+        tuple val(library), path(bam), path(bai), val(barcodes)
+
+    output:
+        tuple val(params.email), val(library), path('*_metrics'), emit: for_agg
+
+    shell:
+    '''
+    # Above params.min_mapq value (default=20)
+    echo "insert_size   2bparsedByRailsModel" > !{library}_insertsize_metrics
+
+    mkfifo good_mapq
+    mkfifo bad_mapq
+    samtools view -hq20 -U bad_mapq !{bam} > good_mapq &
+
+    picard -Xmx16g CollectInsertSizeMetrics --INCLUDE_DUPLICATES --VALIDATION_STRINGENCY SILENT -I good_mapq -O temp.out.txt --MINIMUM_PCT 0 --Histogram_FILE /dev/null &
+    picard -Xmx16g CollectInsertSizeMetrics --INCLUDE_DUPLICATES --VALIDATION_STRINGENCY SILENT -I bad_mapq -O temp2.out.txt --MINIMUM_PCT 0 --Histogram_FILE /dev/null &
+
+    while [[ ! -f "temp.out.txt" || ! -f "temp2.out.txt" ]]; do sleep 2; done
```

Comment:
I changed it to a new version using wait instead but I added hopefully unique names to the pipes to avoid hypothetical situations where these 2 pipes are created by parallel processes (I would believe that nextflow would not allow for that but I don't know). I haven't test yet this!